### ISSUE TITLE
docker builder is just badly designed

### ISSUE THREAD

**Issue by astronouth7303** [cirruslabs/cirrus-ci-docs#427]
(per https://twitter.com/AstraLuma/status/1164377187274940417)

The docker builder is basically travis: Have a vaguely defined pre-configured image, and spend lots and lots and lots of CI cycles even figuring out what you have.

Hopefully what you need is part of it, or hopefully cirrus remembered to include a package manager.

CI collectively moved to container-based builds for good reasons. Really good UI/UX reasons. Docker builder basically abandons that.

----

To tell a concrete story, I was trying to automate building xonsh containers, and was struggling to get xonsh installed. Oddly, I wanted to use the xonsh language to package xonsh.

* No `pip`, no `ensurepip` (which means you removed pip on purpose, which makes installing modern python software far too much effort)
* Don't know what distro this is, so I don't know what package manager to use
* Bind mounting the docker socket into a container lead to some really.... _interesting_ error messages like "error checking context: 'no permission to read from '/proc/1/mem''."

https://github.com/xonsh/container/commit/006ba63fddca2dc73a532ad79d163666c05b3aa9 is the last commit that used a xonsh-based build script.

I know that using docker from inside a container is hard, but y'all dropped the ball on this one.

----

I don't want this bug to be "you need pip on your docker builder image". I want this bug to be "You need to have an option to use docker from a container". I've learned to expect amazing things from Cirrus CI from previous interactions; this is the first time I've been so disappointed and frustrated using your service.

**Comment by fkorotkov** [comment]
It seems you would like something similar to [Docker-in-Docker on GKE](https://cirrus-ci.org/guide/docker-builds-on-kubernetes/) for community containers. Did I understand you correctly?

**Comment by astronouth7303** [comment]
Yeah, not dissimilar? The twitter thread discusses some of the options of how to do it.

To be honest, the delay from spinning up a VM to do it doesn't bother me nearly as much as the "where the hell is this executing and what do I have????" of being dropped into a mysterious linux I don't control.

Something like "spin up a VM, and then spawn a docker pod in it with the build steps" would be perfectly acceptable to me? That makes the fact that docker-enabled jobs need to run in their own VMs an implementation detail that's largely hidden from the user (besides performance)--the user is still mostly using container based jobs, just with a flag to add a docker control socket. The fact that their build job now runs on its own VM instead of on a container fabric is abstracted from the user.

(Of course, I'm only considering the developer that's configuring the pipelines, not the sysadmin that's integrating their infra into cirrus. But, then again, having the abstraction of a container means how precisely it gets executed matters much less, so a variety of strategies can be employed.)

**Comment by fkorotkov** [comment]
> Something like "spin up a VM, and then spawn a docker pod in it with the build steps" would be perfectly acceptable to me? That makes the fact that docker-enabled jobs need to run in their own VMs an implementation detail that's largely hidden from the user (besides performance)--the user is still mostly using container based jobs, just with a flag to add a docker control socket. The fact that their build job now runs on its own VM instead of on a container fabric is abstracted from the user.

Exactly what I had in mind and there is some preliminary work already done as part of #412.

**Comment by astronouth7303** [comment]
Yeah. I think the important thing for me is that I don't care about how or why a job executes under various strategies. I just want to say "Run these commands on this container with a docker socket" and let Cirrus pick the details.

If I wanted to control the details, I'd be using my own GitLab runners or jenkins or fake CI with a webhook receiver. (Especially as both cloud services and github features improve.) I use services like Cirrus because the details don't matter to me.

I want to:
1. Generate Dockerfiles
2. Build images
3. Push those images to a registry

**Comment by RDIL** [comment]
My god Fedor you've been busy today
